File: examples/showcase.fs
// Hatsune Miku Theme - F# Showcase
// All-Miku Synthesis: Every voice, one stage.

namespace MikuTheme.Showcase

open System
open System.Threading.Tasks

// Module with constants
module Constants =
    [<Literal>]
    let CanonicalColor = "#39C5BB"

    [<Literal>]
    let DefaultBpm = 39

    let MaxEnergy = 100

// Discriminated Union (Enum-like)
type MikuVersion =
    | V2Classic
    | V3
    | V4X
    | NT
    | Sekai
    | V6AI

// Discriminated Union with data
type StageEvent =
    | Start of timestamp: DateTime
    | End of duration: TimeSpan
    | Error of message: string
    | Song of title: string * bpm: int

// Record type
type VoiceBank = {
    Name: string
    Version: MikuVersion
    FrequencyRange: int * int
    IsActive: bool
}

// Record with mutable field
type PerformanceState = {
    mutable Energy: int
    mutable CurrentSong: string option
}

// Type alias
type SongProcessor = string -> string
type AsyncSongProcessor = string -> Async<string>

// Interface
type ISynthesizer =
    abstract member Sing: string -> string
    abstract member Version: MikuVersion

// Class with interface
type DigitalDiva(name: string, ?version: MikuVersion) =
    let mutable energy = Constants.MaxEnergy
    let version = defaultArg version V2Classic
    let voiceBank = {
        Name = name
        Version = version
        FrequencyRange = (80, 1100)
        IsActive = true
    }

    // Static member
    static member val ModelId = "CV01" with get

    // Properties
    member _.Name = name
    member _.VoiceBank = voiceBank

    member _.Energy
        with get() = energy
        and set(value) = energy <- max 0 (min Constants.MaxEnergy value)

    // Methods
    member this.Sing(song: string) =
        if energy < 10 then
            failwith "Low energy: Please recharge with leeks"
        energy <- energy - 10
        sprintf "[MIKU] Now singing: %s" song

    // Async method
    member this.SingAsync(song: string) = async {
        do! Async.Sleep 100
        return this.Sing(song)
    }

    // Interface implementation
    interface ISynthesizer with
        member this.Sing(song) = this.Sing(song)
        member _.Version = version

    // Override
    override _.ToString() =
        sprintf "DigitalDiva(%s, %A)" name version

// Module with functions
module MikuUtils =
    // Simple function
    let formatSong title = sprintf "[MIKU] %s" title

    // Function with pattern matching
    let getVersionColor = function
        | V2Classic -> "#39C5BB"
        | NT -> "#00BCD4"
        | Sekai -> "#33CCBB"
        | _ -> "#FFFFFF"

    // Curried function
    let add x y = x + y
    let add10 = add 10

    // Pipeline operators
    let processTitle title =
        title
        |> String.map Char.ToUpper
        |> fun s -> s.Trim()
        |> formatSong

    // Composition operator
    let processAndPrint = formatSong >> printfn "%s"

    // Higher-order function
    let withLogging f x =
        printfn "Input: %A" x
        let result = f x
        printfn "Output: %A" result
        result

    // Partial application
    let loggedFormat = withLogging formatSong

    // Active pattern
    let (|ValidBpm|InvalidBpm|) bpm =
        if bpm >= 60 && bpm <= 240 then ValidBpm bpm
        else InvalidBpm

    // Parameterized active pattern
    let (|DivisibleBy|_|) divisor n =
        if n % divisor = 0 then Some(n / divisor)
        else None

// Pattern matching examples
module Patterns =
    // Match expression
    let handleEvent = function
        | Start ts -> sprintf "Started at %A" ts
        | End duration -> sprintf "Duration: %A" duration
        | Error msg -> sprintf "Error: %s" msg
        | Song (title, bpm) -> sprintf "Playing %s at %d BPM" title bpm

    // Guard clauses
    let categorizeEnergy energy =
        match energy with
        | e when e >= 80 -> "Full power"
        | e when e >= 50 -> "Normal"
        | e when e >= 20 -> "Low"
        | _ -> "Critical"

    // List patterns
    let rec processPlaylist = function
        | [] -> printfn "Playlist complete"
        | [last] -> printfn "Final song: %s" last
        | first :: rest ->
            printfn "Playing: %s" first
            processPlaylist rest

    // Tuple patterns
    let describeRange = function
        | (min, max) when min = max -> "Single frequency"
        | (min, max) when max - min < 100 -> "Narrow range"
        | _ -> "Wide range"

// Computation expressions
module Async =
    // Async workflow
    let performConcert (diva: DigitalDiva) songs = async {
        for song in songs do
            let! result = diva.SingAsync(song)
            printfn "%s" result
            do! Async.Sleep 500
    }

    // Parallel async
    let parallelPerform songs =
        songs
        |> List.map (fun s -> async { return MikuUtils.formatSong s })
        |> Async.Parallel

// Sequences and collections
module Collections =
    // Sequence expression
    let frequencies = seq {
        for i in 80 .. 10 .. 1100 do
            yield i
    }

    // List comprehension
    let squares = [ for x in 1..10 -> x * x ]

    // Array comprehension
    let cubes = [| for x in 1..10 -> x * x * x |]

    // Filtering
    let highFrequencies =
        frequencies
        |> Seq.filter (fun f -> f > 500)
        |> Seq.take 10
        |> Seq.toList

    // Mapping
    let formattedSongs songs =
        songs
        |> List.map MikuUtils.formatSong

    // Folding
    let totalDuration songs =
        songs
        |> List.fold (fun acc (_, duration) -> acc + duration) 0

// Type providers would go here (requires FSharp.Data)
// type JsonProvider = JsonProvider<"sample.json">

// Entry point
[<EntryPoint>]
let main args =
    // Let bindings
    let miku = DigitalDiva("Hatsune Miku", V2Classic)
    let songs = ["Melt"; "World is Mine"; "Love is War"]

    // Mutable binding
    let mutable count = 0

    // Pattern matching in let
    let (min, max) = miku.VoiceBank.FrequencyRange

    // Option handling
    let maybeSong: string option = Some "Rolling Girl"
    match maybeSong with
    | Some s -> printfn "Song: %s" s
    | None -> printfn "No song"

    // Result type
    let tryParseBpm (s: string) =
        match Int32.TryParse(s) with
        | true, n when n >= 60 && n <= 240 -> Ok n
        | true, _ -> Error "BPM out of range"
        | false, _ -> Error "Invalid number"

    // For loop
    for song in songs do
        printfn "%s" (miku.Sing(song))
        count <- count + 1

    // While loop
    while miku.Energy > 50 do
        ignore (miku.Sing("Loop song"))

    // Try-with
    try
        ignore (miku.Sing("Final song"))
    with
    | :? InvalidOperationException as ex ->
        printfn "Operation error: %s" ex.Message
    | ex ->
        printfn "Error: %s" ex.Message

    // Use for IDisposable
    // use reader = new StreamReader("file.txt")

    printfn "Performance complete! Songs: %d" count
    0 // Exit code

File: examples/showcase.dart
/// Hatsune Miku Theme - Dart Showcase
/// All-Miku Synthesis: Every voice, one stage.

import 'dart:async';
import 'dart:collection';
import 'dart:convert';

// Constants
const String canonicalColor = '#39C5BB';
const int defaultBpm = 39;
const maxEnergy = 100;

// Enum with values
enum MikuVersion {
  v2Classic('V2', 2007),
  v3('V3', 2013),
  v4x('V4X', 2016),
  nt('NT', 2020),
  sekai('SEKAI', 2020);

  final String code;
  final int releaseYear;

  const MikuVersion(this.code, this.releaseYear);
}

// Typedef for function types
typedef SongProcessor = String Function(String song);
typedef AsyncSongProcessor = Future<String> Function(String song);

// Mixin
mixin Performable {
  bool _isPerforming = false;

  void startPerformance() {
    _isPerforming = true;
    print('Performance started');
  }

  void endPerformance() {
    _isPerforming = false;
    print('Performance ended');
  }
}

// Abstract class
abstract class Synthesizer {
  String sing(String song);
  MikuVersion get version;

  // Factory constructor in abstract class
  factory Synthesizer.create(String name) = DigitalDiva;
}

// Interface (implicit)
abstract interface class Configurable {
  void configure(Map<String, dynamic> config);
}

// Record class (Dart 3.0+)
class VoiceBank {
  final String name;
  final MikuVersion version;
  final (int, int) frequencyRange;

  const VoiceBank({
    required this.name,
    required this.version,
    this.frequencyRange = (80, 1100),
  });

  // Named constructor
  VoiceBank.classic(String name)
      : this(name: name, version: MikuVersion.v2Classic);

  // Redirecting constructor
  VoiceBank.nt(String name) : this(name: name, version: MikuVersion.nt);
}

// Main class with mixin
class DigitalDiva with Performable implements Synthesizer, Configurable {
  // Static constant
  static const String modelId = 'CV01';

  // Private fields
  final String _name;
  int _energy = maxEnergy;
  late final VoiceBank _voiceBank;

  // Nullable field
  String? _currentSong;

  // Getter
  String get name => _name;
  int get energy => _energy;

  @override
  MikuVersion get version => _voiceBank.version;

  // Setter
  set energy(int value) {
    _energy = value.clamp(0, maxEnergy);
  }

  // Constructor with initializer list
  DigitalDiva(this._name, [MikuVersion version = MikuVersion.v2Classic])
      : _voiceBank = VoiceBank(name: _name, version: version) {
    print('Created $_name');
  }

  // Named constructor
  DigitalDiva.withVoiceBank(VoiceBank voiceBank)
      : _name = voiceBank.name,
        _voiceBank = voiceBank;

  // Factory constructor
  factory DigitalDiva.fromJson(Map<String, dynamic> json) {
    return DigitalDiva(
      json['name'] as String,
      MikuVersion.values.firstWhere(
        (v) => v.code == json['version'],
        orElse: () => MikuVersion.v2Classic,
      ),
    );
  }

  @override
  String sing(String song) {
    if (_energy < 10) {
      throw StateError('Low energy: Please recharge with leeks');
    }
    _currentSong = song;
    _energy -= 10;
    return '[MIKU] Now singing: $song';
  }

  // Async method
  Future<String> singAsync(String song) async {
    await Future.delayed(const Duration(milliseconds: 100));
    return sing(song);
  }

  // Stream generator
  Stream<String> performConcert(List<String> songs) async* {
    for (final song in songs) {
      yield sing(song);
      await Future.delayed(const Duration(milliseconds: 500));
    }
  }

  @override
  void configure(Map<String, dynamic> config) {
    if (config.containsKey('energy')) {
      energy = config['energy'] as int;
    }
  }

  // Operator overloading
  @override
  bool operator ==(Object other) =>
      other is DigitalDiva && other._name == _name && other.version == version;

  @override
  int get hashCode => Object.hash(_name, version);

  // ToString
  @override
  String toString() => 'DigitalDiva($_name, ${version.code})';
}

// Extension methods
extension StringExtension on String {
  String toMikuFormat() => '[MIKU] $this';

  bool get isMikuRelated => toLowerCase().contains('miku');
}

// Generic class
class MetadataStore<T> {
  final Map<String, T> _data = {};

  void set(String key, T value) => _data[key] = value;
  T? get(String key) => _data[key];

  // Generic method
  R transform<R>(String key, R Function(T) transformer) {
    final value = _data[key];
    if (value == null) throw ArgumentError('Key not found: $key');
    return transformer(value);
  }
}

// Sealed class (Dart 3.0+)
sealed class StageEvent {}

class StartEvent extends StageEvent {
  final DateTime timestamp;
  StartEvent(this.timestamp);
}

class EndEvent extends StageEvent {
  final Duration duration;
  EndEvent(this.duration);
}

class ErrorEvent extends StageEvent {
  final String message;
  ErrorEvent(this.message);
}

// Pattern matching (Dart 3.0+)
String handleEvent(StageEvent event) {
  return switch (event) {
    StartEvent(timestamp: var t) => 'Started at $t',
    EndEvent(duration: var d) => 'Duration: ${d.inSeconds}s',
    ErrorEvent(message: var m) => 'Error: $m',
  };
}

// Top-level function
String formatSong(String title, {int? bpm, bool uppercase = false}) {
  final formatted = bpm != null ? '$title @ ${bpm}BPM' : title;
  return uppercase ? formatted.toUpperCase() : formatted;
}

// Main entry point
void main(List<String> args) async {
  // Variable declarations
  var miku = DigitalDiva('Hatsune Miku');
  final songs = ['Melt', 'World is Mine', 'Love is War'];
  const config = {'energy': 100};

  // Null-aware operators
  String? nullableSong;
  final song = nullableSong ?? 'Default Song';
  nullableSong ??= 'Assigned if null';

  // Collection literals
  final list = [1, 2, 3];
  final set = {1, 2, 3};
  final map = {'key': 'value'};

  // Spread operator
  final allSongs = [...songs, 'Rolling Girl'];

  // Collection if/for
  final conditional = [
    'Always',
    if (miku.energy > 50) 'High energy song',
    for (var s in songs) s.toUpperCase(),
  ];

  // Cascade notation
  miku
    ..configure(config)
    ..startPerformance();

  // Try-catch
  try {
    for (final s in songs) {
      print(miku.sing(s));
    }
  } on StateError catch (e) {
    print('State error: ${e.message}');
  } catch (e, stackTrace) {
    print('Error: $e');
    print('Stack: $stackTrace');
  } finally {
    miku.endPerformance();
  }

  // Async/await
  final result = await miku.singAsync('Rolling Girl');
  print(result);

  // Stream handling
  await for (final performance in miku.performConcert(['A', 'B'])) {
    print(performance);
  }

  print('Performance complete!');
}
